### PR TITLE
[lldb][ClangExpressionParser] Set BuiltinHeadersInSystemModules depending on SDK version

### PR DESCRIPTION
**Pull request by Michael137**
Abandoned https://github.com/llvm/llvm-project/pull/101778 in favour of this

This patch changes the way we initialize `BuiltinHeadersInSystemModules` which is one of the flags controlling Clang's behaviour when the Darwin module is split into more fine-grained modules. The ClangExpressionParser currently unconditionally sets `-fbuiltin-headers-in-system-modules` when evaluating expressions with the `target.import-std-module` setting. This flag should, however, be set depending on the SDK version (which is what the Clang Darwin toolchain does).

Unfortunately, the compiler instance that we create with `ClangExpressionParser` never consults the Clang driver, and thus doesn't correctly infer `BuiltinHeadersInSystemModules`. Note, this isn't an issue with the `CompilerInstance` that the `ClangModulesDeclVendor` creates because it uses the `createInovcation` API, which calls into `Darwin::addClangTargetOptions`.

This patch mimicks how `sdkSupportsBuiltinModules` is used in `Darwin::addClangTargetOptions`.

This ensures that the `import-std-module` API tests run cleanly regardless of SDK version.

The plan is to eventually make the `CompilerInstance` construction in `ClangExpressionParser` go through the driver, so we can avoid duplicating the logic in LLDB. But we aren't there yet.

**Implementation**
* We look for the `SDKSettings.json` in the sysroot directory that we found in DWARF (via `DW_AT_LLVM_sysroot`)
* Then parse this file and extract the SDK version number out of it
* Then mimick `sdkSupportsBuiltinModules` from `Toolchains/Darwin.cpp` and set `BuiltinHeadersInSystemModules` based on it

rdar://116490281

### CODE REVIEW

**Comment by Michael137** [comment]
@swift-ci test

File: lldb/source/Plugins/ExpressionParser/Clang/ClangExpressionParser.cpp [hunk 1]
```diff
@@ -352,6 +353,73 @@ static void SetupDefaultClangDiagnostics(CompilerInstance &compiler) {
   }
 }
 
+// NOTE: should be kept in sync with sdkSupportsBuiltinModules in
+// Toolchains/Darwin.cpp
+static bool
+sdkSupportsBuiltinModulesImpl(const llvm::Triple &triple,
```
Comment (ian-twilightcoder) [inline lldb/source/Plugins/ExpressionParser/Clang/ClangExpressionParser.cpp:359]: I'm about to change the implementation of this function in the Darwin driver

**Comment by ian-twilightcoder** [comment]
@cachemeifyoucan @Bigcheese @cyndyishida is it really that difficult to change this usage to use `clang::createInvocation` or `clang::CompilerInvocation::CreateFromArgs` or `clang::BuildCompilation`?

**Comment by Michael137** [comment]
> @cachemeifyoucan @Bigcheese @cyndyishida is it really that difficult to change this usage to use `clang::createInvocation` or `clang::CompilerInvocation::CreateFromArgs` or `clang::BuildCompilation`?

I have something that works to some extent. But it's definitely a risky enough change to land for 6.0

File: lldb/source/Plugins/ExpressionParser/Clang/ClangExpressionParser.cpp (hunk 1 above)
Comment (Michael137) [reply lldb/source/Plugins/ExpressionParser/Clang/ClangExpressionParser.cpp:359]: Implementation in what sense? Changes to the SDK version checks?

**Comment by cachemeifyoucan** [comment]
> is it really that difficult to change this usage to use `clang::createInvocation` or `clang::CompilerInvocation::CreateFromArgs` or `clang::BuildCompilation`?

I have been advocating lldb to change to use them to avoid a lot of the module configuration issues.

File: lldb/source/Plugins/ExpressionParser/Clang/ClangExpressionParser.cpp (hunk 1 above)
Comment (ian-twilightcoder) [reply lldb/source/Plugins/ExpressionParser/Clang/ClangExpressionParser.cpp:359]: Posted https://github.com/llvm/llvm-project/pull/102239 for it.

**Comment by cyndyishida** [comment]
> > @cachemeifyoucan @Bigcheese @cyndyishida is it really that difficult to change this usage to use `clang::createInvocation` or `clang::CompilerInvocation::CreateFromArgs` or `clang::BuildCompilation`?
> 
> I have something that works to some extent. But it's definitely a risky enough change to land for 6.0

Whats the risk involved?

File: lldb/source/Plugins/ExpressionParser/Clang/ClangExpressionParser.cpp [hunk 2]
```diff
@@ -576,7 +644,8 @@ ClangExpressionParser::ClangExpressionParser(
     lang_opts.GNUMode = true;
     lang_opts.GNUKeywords = true;
     lang_opts.CPlusPlus11 = true;
-    lang_opts.BuiltinHeadersInSystemModules = true;
+    lang_opts.BuiltinHeadersInSystemModules = !sdkSupportsBuiltinModules(
```
Comment (adrian-prantl) [inline lldb/source/Plugins/ExpressionParser/Clang/ClangExpressionParser.cpp:647]: This `!` is the intended direction, right?

**Review by adrian-prantl — APPROVED** [review]
LGTM as a surgical fix for the `swift/release/6.0` branch.

**Comment by adrian-prantl** [comment]
> > > @cachemeifyoucan @Bigcheese @cyndyishida is it really that difficult to change this usage to use `clang::createInvocation` or `clang::CompilerInvocation::CreateFromArgs` or `clang::BuildCompilation`?
> > 
> > 
> > I have something that works to some extent. But it's definitely a risky enough change to land for 6.0
> 
> Whats the risk involved?

I wouldn't want to change the way we setup up the CompilerInvocation this late in the release branch's life.

File: lldb/source/Plugins/ExpressionParser/Clang/ClangExpressionParser.cpp (hunk 2 above)
Comment (ian-twilightcoder) [reply lldb/source/Plugins/ExpressionParser/Clang/ClangExpressionParser.cpp:647]: Yes. The builtin headers have to go in the system modules if the sdk doesn't support the builtin modules.

File: lldb/source/Plugins/ExpressionParser/Clang/ClangExpressionParser.cpp (hunk 1 above)
Comment (Michael137) [reply lldb/source/Plugins/ExpressionParser/Clang/ClangExpressionParser.cpp:359]: Ah yes I accounted for that in this patch already

**Pull request merged**